desc/objectives/_coils.py

```diff
@@ -602,11 +603,154 @@ def compute(self, params, constants=None):
+        Collocation grid used to discritize each coil. Default = LinearGrid(N=16)
+    name : str, optional
+        Name of the objective function.
+
+    """
+
+    _scalar = False
+    _units = "(m)"
+    _print_value_fmt = "Minimum coil-coil distance: {:10.3e} "
+
+    def __init__(
+        self,
+        coils,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        grid=None,
+        name="coilset minimum distance",
+    ):
+        if target is None and bounds is None:
+            bounds = (0, np.inf)
+        self._grid = grid
+        super().__init__(
+            things=coils,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        self._coilset = self.things[0]
+        if self._grid is None:
+            self._grid = LinearGrid(N=16)
+        else:
+            self._grid = self._grid
+
+        num_nodes = self._grid.num_nodes
+        perm_idx = np.tile(np.arange(num_nodes), 2)
+        self._perms = [perm_idx[k : k + num_nodes] for k in range(num_nodes)]
+
+        self._dim_f = self._coilset.num_coils
+
+        if self._normalize:
+            coils = tree_leaves(
+                self._coilset, is_leaf=lambda x: not hasattr(x, "__len__")
+            )
+            scales = [compute_scaling_factors(coil)["a"] for coil in coils]
+            self._normalization = np.mean(scales)  # mean length of coils
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params, constants=None):
+        """Compute minimum distances between coils.
+
+        Parameters
+        ----------
+        params : dict
+            Dictionary of coilset degrees of freedom, eg CoilSet.params_dict
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc.
+            Defaults to self._constants.
+
+        Returns
+        -------
+        f : array of floats
+            Minimum distance to another coil for each coil in the coilset.
+
+        """
+        # include all permutations of grid nodes between coils
+        xyz = self._coilset.compute_position(source_grid=self._grid)
+        pts1 = jnp.tile(xyz, (1, len(self._perms), 1))
+        pts2 = jnp.concatenate([xyz[:, p, :] for p in self._perms], axis=1)
+
+        def body(k):
+            dist = safenorm(pts1[k] - pts2, axis=-1)  # distances between pts
```

Comment:
instead of doing all the permutation stuff I think we can just broadcast these differently. Assuming `pts` has shape `(num_coils, num_nodes, 3)`, we want `pts[i][None,:,None] - pts[:,None,:]` so that `dist` has shape `(num_coils, num_nodes, num_nodes)`